packages/react-devtools-scheduling-profiler/src/view-base/View.js

```diff
@@ -0,0 +1,274 @@
+    this.subviews = [];
+  }
+
+  /**
+   * Executes the display flow if this view needs to be drawn.
+   *
+   * 1. Lays out subviews with `layoutSubviews`.
+   * 2. Draws content with `draw`.
+   */
+  displayIfNeeded(context: CanvasRenderingContext2D) {
+    if (
+      (this._needsDisplay || this._subviewsNeedDisplay) &&
+      rectIntersectsRect(this.frame, this.visibleArea) &&
+      !sizeIsEmpty(this.visibleArea.size)
+    ) {
+      this.layoutSubviews();
+      if (this._needsDisplay) this._needsDisplay = false;
+      if (this._subviewsNeedDisplay) this._subviewsNeedDisplay = false;
+      this.draw(context);
+    }
+  }
+
+  /**
+   * Layout self and subviews.
+   *
+   * Implementations should call `setNeedsDisplay` if a draw is required.
+   *
+   * The default implementation uses `this.layouter` to lay out subviews.
+   *
+   * Can be overwritten by subclasses that wish to manually manage their
+   * subviews' layout.
+   *
+   * NOTE: Do not call directly! Use `displayIfNeeded`.
+   *
+   * @see displayIfNeeded
+   */
+  layoutSubviews() {
+    const {frame, _layouter, subviews, visibleArea} = this;
+    const existingLayout = viewsToLayout(subviews);
+    const newLayout = _layouter(existingLayout, frame);
+    collapseLayoutIntoViews(newLayout);
+
+    subviews.forEach((subview, subviewIndex) => {
+      if (rectIntersectsRect(visibleArea, subview.frame)) {
+        subview.setVisibleArea(intersectionOfRects(visibleArea, subview.frame));
+      } else {
+        subview.setVisibleArea(zeroRect);
+      }
+    });
+  }
+
+  /**
+   * Draw the contents of this view in the given canvas `context`.
+   *
+   * Defaults to drawing this view's `subviews`.
+   *
+   * To be overwritten by subclasses that wish to draw custom content.
+   *
+   * NOTE: Do not call directly! Use `displayIfNeeded`.
+   *
+   * @see displayIfNeeded
+   */
+  draw(context: CanvasRenderingContext2D) {
+    const {subviews, visibleArea} = this;
+    subviews.forEach(subview => {
+      if (rectIntersectsRect(visibleArea, subview.visibleArea)) {
+        subview.displayIfNeeded(context);
+      }
+    });
+  }
+
+  /**
+   * Handle an `interaction`.
+   *
+   * To be overwritten by subclasses that wish to handle interactions.
+   */
+  // Internal note: Do not call directly! Use
+  // `handleInteractionAndPropagateToSubviews` so that interactions are
+  // propagated to subviews.
+  handleInteraction(interaction: Interaction) {}
+
+  /**
+   * Handle an `interaction` and propagates it to all of this view's
+   * `subviews`.
+   *
+   * NOTE: Should not be overridden! Subclasses should override
+   * `handleInteraction` instead.
+   *
+   * @see handleInteraction
+   * @protected
+   */
+  handleInteractionAndPropagateToSubviews(interaction: Interaction) {
+    this.handleInteraction(interaction);
+    this.subviews.forEach(subview =>
+      subview.handleInteractionAndPropagateToSubviews(interaction),
+    );
+  }
```

Comment:
Hmm, maybe I'm not understanding your `ViewManager` design correctly. If the `ViewManager` class is an implementation detail of the `View` class, won't `View` still need to implement all its existing methods, including `handleInteractionAndPropagateToSubviews`?

What about this alternative design:

1. Turn `View` into an interface that only implements e.g. `draw`, `handleInteraction`, etc.
1. Define a new *public* `ViewManager` class that handles the invalidation and interaction propagation logic. 

`CanvasPage` will then instantiate `ViewManager`s and inject concrete `View` implementations into them.

This will mean that the manager will handle subViewManager/superViewManager relationships. Also, `View` implementations that instantiate their own subviews (e.g. `ResizableSplitView`, `FlamechartView`) will need to know how to instantiate `ViewManager`s, which doesn't feel right.